File: backend/app/api/v1/tags.py
"""
标签管理 API：列表、创建、素材绑定标签
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from typing import List, Optional
from sqlalchemy.orm import Session
from sqlalchemy import select, func

from app.api import deps
from app.models.tag import Tag, material_tags
from app.utils.response import success

router = APIRouter(prefix="/tags", tags=["标签"])


class TagCreate(BaseModel):
    name: str


class TagOut(BaseModel):
    id: int
    name: str
    usage_count: int

    class Config:
        from_attributes = True


class MaterialTagsUpdate(BaseModel):
    tag_ids: List[int]


@router.get("")
def list_tags(
    q: str = Query("", description="按名称搜索"),
    db: Session = Depends(deps.get_db),
):
    """获取标签列表，支持按名称搜索"""
    stmt = select(Tag).order_by(Tag.usage_count.desc(), Tag.id.asc())
    if q.strip():
        stmt = stmt.where(Tag.name.contains(q.strip()))
    tags = db.execute(stmt).scalars().all()
    data = [{"id": t.id, "name": t.name, "usage_count": t.usage_count} for t in tags]
    return success(data=data)


@router.post("")
def create_tag(payload: TagCreate, db: Session = Depends(deps.get_db)):
    """创建标签（名称唯一）"""
    name = payload.name.strip()
    if not name:
        raise HTTPException(status_code=400, detail="标签名称不能为空")
    existing = db.execute(select(Tag).where(Tag.name == name)).scalars().first()
    if existing:
        raise HTTPException(status_code=400, detail="该标签已存在")
    tag = Tag(name=name, usage_count=0)
    db.add(tag)
    db.commit()
    db.refresh(tag)
    return success(data={"id": tag.id, "name": tag.name, "usage_count": tag.usage_count})


@router.get("/{tag_id}")
def get_tag(tag_id: int, db: Session = Depends(deps.get_db)):
    """获取单个标签"""
    tag = db.get(Tag, tag_id)
    if not tag:
        raise HTTPException(status_code=404, detail="标签不存在")
    return success(data={"id": tag.id, "name": tag.name, "usage_count": tag.usage_count})


@router.delete("/{tag_id}")
def delete_tag(tag_id: int, db: Session = Depends(deps.get_db)):
    """删除标签（会解除与素材的关联）"""
    tag = db.get(Tag, tag_id)
    if not tag:
        raise HTTPException(status_code=404, detail="标签不存在")
    db.delete(tag)
    db.commit()
    return success(message="删除成功")

File: frontend/src/api/tags.js
import request from "./index";

/**
 * 获取标签列表（支持按名称搜索）
 * @param {Object} [params] - 查询参数
 * @param {string} [params.q] - 搜索关键词
 * @returns {Promise} 标签列表
 */
export function fetchTags(params) {
  return request.get("/api/v1/tags", { params });
}

/**
 * 创建标签
 * @param {Object} data - 标签数据
 * @param {string} data.name - 标签名称
 * @returns {Promise} 创建的标签
 */
export function createTag(data) {
  return request.post("/api/v1/tags", data);
}

/**
 * 获取单个标签
 * @param {number} tagId - 标签 ID
 * @returns {Promise} 标签详情
 */
export function getTag(tagId) {
  return request.get(`/api/v1/tags/${tagId}`);
}

/**
 * 删除标签
 * @param {number} tagId - 标签 ID
 * @returns {Promise} 删除结果
 */
export function deleteTag(tagId) {
  return request.delete(`/api/v1/tags/${tagId}`);
}
